PubNub/Data/Managers/PNTelemetry.m

```diff
@@ -0,0 +1,296 @@
+
+/**
+ @brief  Stores reference on key under which request status is stored.
+ 
+ @since 4.6.2
+ */
+static NSString * const kPNOperationDateKey = @"d";
+
+/**
+ @brief  Stores reference on key under which service advisory information stored.
+ 
+ @since 4.6.2
+ */
+static NSString * const kPNOperationLatencyKey = @"l";
+
+
+NS_ASSUME_NONNULL_BEGIN
+
+#pragma mark - Protected interface declaration
+
+@interface PNTelemetry ()
+
+
+#pragma mark - Information 
+
+/**
+ @brief  Stores reference on list of
+ */
+@property (nonatomic, strong) NSMutableDictionary<NSString *, NSMutableArray<NSDictionary *> *> *latencies;
+/**
+ @brief      Stores reference on map of active operations and their start date.
+ @discussion Dictionary used by start/stop latency measure API to calculate operation latency from time which
+             has been required to receive response (if metrics not available).
+ 
+ @since 4.6.2
+ */
+@property (nonatomic, strong) NSMutableDictionary *trackedLatencies;
+
+/**
+ @brief      Stores reference on timer which is used to clean accumulated latencies information.
+ @discussion SDK required to store latency information for last usage minute, so all outdated latencies 
+             information should be removed from persistent storage.
+ 
+ @since 4.6.2
+ */
+@property (nonatomic, strong, nullable) NSTimer *cleanUpTimer;
+
+/**
+ @brief  Spin/unfair-lock which is used to protect access to shared resources from multiple threads.
+ 
+ @since 4.6.2
+ */
+@property (nonatomic, assign) os_unfair_lock resourceAccessLock;
+
+
+#pragma mark - Operation information
+
+/**
+ @brief      Shortened name of API endpoint for specific operation.
+ @discussion Some operations refer to single endpoint with only difference in passed parameters, but actual
+             endpoint is the same. This method return shortened name of this endpoint.
+ 
+ @param operationType One of \b PNOperationType enumerator fields which describe for which operation endpoint
+                      should be retrieved.
+ 
+ @return Shortened API endpoint name.
+ 
+ @since 4.6.2
+ */
+- (NSString *)endpointNameForOperation:(PNOperationType)operationType;
+
+
+#pragma mark - Handlers
+
+/**
+ @brief      Handler clean up timmer triggered.
+ @discussion Use this handler to clean up outdated latencies information.
+ 
+ @param timer Reference on timer which triggered callback.
+ 
+ @since 4.6.2
+ */
+- (void)handleCleanUpTimer:(NSTimer *)timer;
+
+#pragma mark -
+
+
+@end
+
+NS_ASSUME_NONNULL_END
+
+
+
+@implementation PNTelemetry
+
+
+#pragma mark - Initialization and configuration
+
+- (instancetype)init {
+    
+    // Cjeck whether initialization has been successful or not.
+    if ((self = [super init])) {
+        
+        self.resourceAccessLock = OS_UNFAIR_LOCK_INIT;
+        self.latencies = [NSMutableDictionary new];
+        self.trackedLatencies = [NSMutableDictionary new];
+        self.cleanUpTimer = [NSTimer scheduledTimerWithTimeInterval:1.0f target:self
```

Comment:
Ah that's a really good point. What if we made it async on a separate latency queue? Is that just overkill and extra complicated?

I usually try to avoid timers, but I hadn't considered it would affect the request/response timing